{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare-kreport-taxonomic-profiles.ipynb\n",
    "\n",
    "# The goal of this Jupyter notebook is to facilitate easy comparisons of the taxonomic \n",
    "# profiles of multiple metagenomic samples. These analyses require kraken-report \n",
    "# (kreport) format files as inputs, and the analyses can be performed for any \n",
    "# taxonomic rank in the kreport files (strain, species, genus, family, etc.). \n",
    "\n",
    "# Running the notebook requires the pandas, seaborn, and matplotlib Python libraries.\n",
    "\n",
    "# The kreport format is standard for many profilers (Kraken, Bracken, Centrifuge, MMSeqs2), and \n",
    "# there are conversion scripts available for Metamaps, MetaPhlAn3, and MEGAN in:\n",
    "# https://github.com/PacificBiosciences/pb-metagenomics-tools/tree/master/pb-metagenomics-scripts\n",
    "# The PacBio MEGAN-LR workflows also produce kreport output formats:\n",
    "# https://github.com/PacificBiosciences/pb-metagenomics-tools/\n",
    "\n",
    "# To run your own analyses, you will need to edit the sample/file path information. \n",
    "# You can generate counts at different taxnomic levels using the get_all_dfs() function.\n",
    "# There are examples of how to do this at the species, genus, and family levels, along with\n",
    "# how to save the count tables and create/save stacked barplot figures. \n",
    "\n",
    "# Daniel Portik\n",
    "# Bioinformatics Scientist, Pacific Biosciences\n",
    "# dportik@pacificbiosciences.com\n",
    "# 05/2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Enter sample and file information here\n",
    "########################################################################\n",
    "\n",
    "# Provide the full path to each kreport to include below. \n",
    "# In this example we are using four files. The variable name\n",
    "# is ideally what you want to call the sample in the analysis.\n",
    "\n",
    "CID_138274 = '/Users/dportik/Documents/Projects/Datasets/CID_138274.megan-RMA-c2c.kreport.txt'\n",
    "CID_138390 = '/Users/dportik/Documents/Projects/Datasets/CID_138390.megan-RMA-c2c.kreport.txt'\n",
    "CID_139369 = '/Users/dportik/Documents/Projects/Datasets/CID_139369.megan-RMA-c2c.kreport.txt'\n",
    "CID_139445 = '/Users/dportik/Documents/Projects/Datasets/CID_139445.megan-RMA-c2c.kreport.txt'\n",
    "\n",
    "# Create a dictionary with string labels for the files as keys and the \n",
    "# file path variables above as the values. The string labels will be used \n",
    "# to label the samples in the resulting tables and figures.\n",
    "\n",
    "# Follow the format below for your own data.\n",
    "\n",
    "file_dict = {\"CID_138274\": CID_138274, \n",
    "             \"CID_138390\": CID_138390, \n",
    "             \"CID_139369\": CID_139369, \n",
    "             \"CID_139445\": CID_139445}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The series of functions will create three types of pandas dataframes with taxon \n",
    "# counts for a specified taxonomic rank. The first dataframe contains absolute counts of \n",
    "# the taxa across all samples. It can be exported as a CSV file, and is also transformed to \n",
    "# make the second and third dataframes. These additional dataframes will contain the \n",
    "# absolute or normalized counts in a transposed format that is used to make the stacked\n",
    "# barplot figures. Do not edit these functions!\n",
    "\n",
    "def get_unique_taxa_at_rank(file_dict, rank):\n",
    "    \"\"\"\n",
    "    Function to obtain all unique taxon names at a given rank \n",
    "    for a set of kreport files provided in file_dict. Returns \n",
    "    sorted list of unique taxon names.\n",
    "    \n",
    "    :param file_dict: dictionary with string label as key and file path as val\n",
    "    :param rank: desired taxonomic rank (e.g., 'S1', 'S', 'G', 'F', etc.)\n",
    "    :return: sorted list of unique taxon names at rank\n",
    "    \"\"\"\n",
    "    # initiate empty set for taxon names\n",
    "    taxon_set = set()\n",
    "    \n",
    "    # iterate over the file dictionary\n",
    "    for key, val in file_dict.items():\n",
    "        \n",
    "        # create a pandas dataframe, remove whitespace to clean names column of kreport \n",
    "        df = pd.read_csv(val, sep = '\\t', header=None, skipinitialspace=True,\n",
    "                         names = ['proportion', 'cumulative_count', 'level_count', 'rank', 'taxid', 'name'])\n",
    "        \n",
    "        # get all unique taxon names at rank, add to set\n",
    "        taxa = [t for t in (df.loc[df['rank'] == rank, 'name']).unique()]\n",
    "        taxon_set.update(taxa)\n",
    "    \n",
    "    # return sorted list of taxon names\n",
    "    return sorted(taxon_set)\n",
    "\n",
    "def make_rank_dataframe(file_dict, rank):\n",
    "    \"\"\"\n",
    "    Function to create a dataframe with taxon counts for \n",
    "    all kreport files in file_dict at a given taxonomic rank. \n",
    "    Returns pandas dataframe.\n",
    "    \n",
    "    :param file_dict: dictionary with string label as key and file path as val\n",
    "    :param rank: desired taxonomic rank (e.g., 'S1', 'S', 'G', 'F', etc.)\n",
    "    :return: pandas dataframe of taxon counts for specified rank\n",
    "    \"\"\"\n",
    "    \n",
    "    # initiate empty dictionary\n",
    "    taxon_count_dict = {}\n",
    "    \n",
    "    # get taxon names to use for this rank\n",
    "    taxon_list = get_unique_taxa_at_rank(file_dict, rank)\n",
    "    \n",
    "    # iterate over the file dictionary\n",
    "    for key, val in file_dict.items():\n",
    "        \n",
    "        # add this sample as a key in the taxon count dictionary\n",
    "        # value is an empty list that will be filled with taxon counts\n",
    "        # in the order of the taxon_list\n",
    "        taxon_count_dict[key] = []\n",
    "        \n",
    "        # make temp dataframe, remove whitespace to clean names column of kreport \n",
    "        df = pd.read_csv(val, sep = '\\t', header=None, skipinitialspace=True, \n",
    "                         names = ['proportion', 'cumulative_count', 'level_count', 'rank', 'taxid', 'name'])\n",
    "        \n",
    "        # iterate over taxon names \n",
    "        for taxon in taxon_list:\n",
    "            # make a subframe with matched name\n",
    "            temp = df.loc[df['name'] == taxon]\n",
    "            # add the cumulative count for the taxon name to the dict list\n",
    "            # this should only be a single name, taking sum as easy way to get integer\n",
    "            taxon_count_dict[key].append(temp.loc[df['rank'] == rank, 'cumulative_count'].sum())\n",
    "    \n",
    "    # create new dataframe from the dictionaries constructed\n",
    "    df_taxa = pd.DataFrame.from_dict(taxon_count_dict)\n",
    "    \n",
    "    # add in a column with the taxon labels\n",
    "    df_taxa.insert(0, 'Taxa', taxon_list)\n",
    "    \n",
    "    # return the taxon dataframe\n",
    "    return df_taxa\n",
    "\n",
    "def prepare_plotting_dfs(df, file_dict, sortby=list(file_dict.keys())):\n",
    "    # transpose dataframe, sort counts, and add new column names\n",
    "    df_transpose = df.sort_values(by=sortby, ascending=False).set_index('Taxa').transpose().rename_axis('Samples', axis=1)\n",
    "    # normalize the species counts of each sample 100%\n",
    "    df_transpose_norm = df_transpose.apply(lambda x: x*100/sum(x), axis=1)\n",
    "    \n",
    "    return df_transpose, df_transpose_norm\n",
    "\n",
    "def get_all_dfs(file_dict, rank, sortby=list(file_dict.keys())):\n",
    "    df_one = make_rank_dataframe(file_dict, rank)\n",
    "    df_two, df_three = prepare_plotting_dfs(df_one, file_dict, sortby=sortby)\n",
    "    return df_one, df_two, df_three\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Create a taxon dataframe for the SPECIES level\n",
    "########################################################################\n",
    "\n",
    "# This demonstrates how to use the main function get_all_dfs() to create \n",
    "# the comparison dataframes. We will start the demonstration for the \n",
    "# species level.\n",
    "\n",
    "# To generate the dataframes, we use the following function: \n",
    "# get_all_dfs(file_dict, rank, sortby='Name')\n",
    "#    the first positional argument is file_dict (leave unchanged)\n",
    "#    the second positional argument is the rank desired ('S1', 'S', 'G', 'F', etc.)\n",
    "#    the optional argument sortby can be used to specify one of the samples to sort \n",
    "#        abundances by, which must be a key name in file_dict\n",
    "# The function returns three dataframes, so three variable names must be specified\n",
    "\n",
    "# Here is the basic usage to get dataframes for species level taxa:\n",
    "df_sp, df_sp_plot, df_sp_plot_norm = get_all_dfs(file_dict, 'S')\n",
    "# In the above example, we labeled the three dataframes returned by the function as:\n",
    "# df_sp, df_sp_plot, and df_sp_plot_norm.\n",
    "\n",
    "# Below is the same example but with sorting by abundances in sample \"CID_139445\", un-hash to run:\n",
    "# df_sp, df_sp_plot, df_sp_plot_norm = get_all_dfs(file_dict, 'S', sortby=\"CID_139445\")\n",
    "# Make sure if you use the optional argument that you specify a label name/key in the file_dict."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preview the first type of dataframe. Here, samples are the columns and taxa form the rows. \n",
    "df_sp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preview the second dataframe. Notice it is a transposed version of the first one.\n",
    "# Here, taxa are columns and samples are the rows.\n",
    "df_sp_plot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preview the third dataframe. This is the same format as above,\n",
    "# but the values have been normalized so counts for each sample \n",
    "# total to 100. \n",
    "df_sp_plot_norm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Saving the dataframes as CSV files\n",
    "########################################################################\n",
    "\n",
    "# to save any of the above dataframes, use the df.to_csv() method\n",
    "\n",
    "# for the df_sp, we don't want to include the index (which is just integers)\n",
    "# we can also specify tab-delimited using the sep argument:\n",
    "df_sp.to_csv(\"Species-Absolute-Abundance-Counts.txt\", index=False, sep='\\t')\n",
    "\n",
    "# for the df_sp_plot and df_sp_plot_norm dataframes, keep the index (which is the sample names):\n",
    "df_sp_plot.to_csv(\"Species-Absolute-Abundance-Counts-Transposed.txt\", sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Make a HORIZONTAL stacked barplot using the NORMALIZED COUNT dataframe.\n",
    "########################################################################\n",
    "\n",
    "# Here you will need to potentially change some of the argument values \n",
    "# depending on the number of samples and taxa in your dataset.\n",
    "\n",
    "# To make plotting easier we can set the color palette here using seaborn. \n",
    "pal = sns.color_palette(\"tab20\")\n",
    "\n",
    "# Create the horizontal stacked bar plot using the pandas .plot.barh() function with the following args:\n",
    "\n",
    "# stacked: controls whether or not to create a stacked barplot\n",
    "# figsize: the dimensions of the figure, in (width, height)\n",
    "# width: this controls the width of the bars, 1 = no separation between them, <1 puts spacing in\n",
    "# color: sets the colors, we refer to pal which was set above\n",
    "# fontsize: the size of sample labels (values = xx-small, x-small, small, medium, large, x-large, xx-large)\n",
    "# edgecolor: controls the color of lines separating colors within bars, as well as bar outline\n",
    "# linewidth: the line size for the edgecolor\n",
    "\n",
    "ax = df_sp_plot_norm.plot.barh(stacked=True, figsize=(20,5), width=0.8, \n",
    "                               color=pal, fontsize='large', edgecolor='black', \n",
    "                               linewidth=0.5)\n",
    "\n",
    "# these commands eliminate the bounding box for the barplot\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['left'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# this controls the placement of the legend (bbox_to_anchor), as well as font (fontsize), \n",
    "# number of columns for labels (ncol), and spacing (labelspacing)\n",
    "ax.legend(bbox_to_anchor=(1,-0.10), fontsize='x-small', ncol=7, labelspacing=0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can save the above plot using the following:\n",
    "ax.figure.savefig('Species-Abundances-Normalized-Horizontal.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Make a VERTICAL stacked barplot using the NORMALIZED COUNT dataframe.\n",
    "########################################################################\n",
    "\n",
    "# Here you will need to potentially change some of the argument values \n",
    "# depending on the number of samples and taxa in your dataset.\n",
    "\n",
    "# To make plotting easier we can set the color palette here using seaborn. \n",
    "pal = sns.color_palette(\"tab20\")\n",
    "\n",
    "# Create the vertical stacked bar plot using the pandas .plot.bar() function with the following args:\n",
    "\n",
    "# stacked: controls whether or not to create a stacked barplot\n",
    "# figsize: the dimensions of the figure, in (width, height)\n",
    "# width: this controls the width of the bars, 1 = no separation between them, <1 puts spacing in\n",
    "# color: sets the colors, we refer to pal which was set above\n",
    "# fontsize: the size of sample labels (values = xx-small, x-small, small, medium, large, x-large, xx-large)\n",
    "# edgecolor: controls the color of lines separating colors within bars, as well as bar outline\n",
    "# linewidth: the line size for the edgecolor\n",
    "\n",
    "ax = df_sp_plot_norm.plot.bar(stacked=True, figsize=(5,15), width=0.8, \n",
    "                               color=pal, fontsize='large', edgecolor='black', \n",
    "                               linewidth=0.5)\n",
    "\n",
    "# these commands eliminate the bounding box for the barplot\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# for vertical plot, we need to reverse the legend labels, start by getting these objects\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "\n",
    "# and add reversed(handles) and reversed(labels) to the legend arguments\n",
    "# this controls the placement of the legend (bbox_to_anchor), as well as font (fontsize), \n",
    "# number of columns for labels (ncol), and spacing (labelspacing)\n",
    "ax.legend(reversed(handles), reversed(labels), bbox_to_anchor=(1,0.9), \n",
    "          fontsize='x-small', ncol=1, labelspacing=0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can save the above plot using the following:\n",
    "ax.figure.savefig('Species-Abundances-Normalized-Vertical.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Make a HORIZONTAL stacked barplot using the ABSOLUTE COUNT dataframe.\n",
    "########################################################################\n",
    "\n",
    "# In this plot, we don't expect the stacked bars to be the same lengths.\n",
    "\n",
    "# All the same arguments are used, but we use the df_sp_plot dataframe instead\n",
    "# of the df_sp_plot_norm dataframe\n",
    "\n",
    "pal = sns.color_palette(\"tab20\")\n",
    "ax = df_sp_plot.plot.barh(stacked=True, figsize=(20,5), width=0.8, \n",
    "                               color=pal, fontsize='large', edgecolor='black', \n",
    "                               linewidth=0.5)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['left'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.legend(bbox_to_anchor=(1,-0.10), fontsize='x-small', ncol=7, labelspacing=0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can save the above plot using the following:\n",
    "ax.figure.savefig('Species-Abundances-Absolute-Horizontal.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Create a taxon dataframe for the GENUS level\n",
    "########################################################################\n",
    "\n",
    "# Do the same thing, but for the genus level\n",
    "\n",
    "# here is the basic usage to get dataframes for genus level taxa\n",
    "df_gn, df_gn_plot, df_gn_plot_norm = get_all_dfs(file_dict, 'G')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preview the first dataframe. Here, samples are the columns and taxa form the rows.\n",
    "df_gn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save genus counts to csv\n",
    "df_gn.to_csv(\"Genus-Absolute-Abundance-Counts.txt\", index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Make a HORIZONTAL stacked barplot using the NORMALIZED COUNT dataframe.\n",
    "########################################################################\n",
    "\n",
    "# make sure to change dataframe name to plot!\n",
    "\n",
    "pal = sns.color_palette(\"tab20\")\n",
    "\n",
    "ax = df_gn_plot_norm.plot.barh(stacked=True, figsize=(20,5), width=0.8, \n",
    "                               color=pal, fontsize='large', edgecolor='black', \n",
    "                               linewidth=0.5)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['left'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.legend(bbox_to_anchor=(0.9,-0.10), fontsize='small', ncol=8, labelspacing=0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Create a taxon dataframe for the FAMILY level\n",
    "########################################################################\n",
    "\n",
    "# Do the same thing, but for the family level\n",
    "\n",
    "# here is the basic usage to get dataframes for genus level taxa\n",
    "df_fm, df_fm_plot, df_fm_plot_norm = get_all_dfs(file_dict, 'F')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preview the first dataframe. Here, samples are the columns and taxa form the rows.\n",
    "df_fm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save family counts to csv\n",
    "df_fm.to_csv(\"Family-Absolute-Abundance-Counts.txt\", index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Make a HORIZONTAL stacked barplot using the NORMALIZED COUNT dataframe.\n",
    "########################################################################\n",
    "\n",
    "# make sure to change dataframe name to plot!\n",
    "\n",
    "pal = sns.color_palette(\"tab20\")\n",
    "\n",
    "ax = df_fm_plot_norm.plot.barh(stacked=True, figsize=(20,5), width=0.8, \n",
    "                               color=pal, fontsize='large', edgecolor='black', \n",
    "                               linewidth=0.5)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['left'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.legend(bbox_to_anchor=(0.8,-0.10), fontsize='small', ncol=7, labelspacing=0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
